File: packages/oxlint-plugin-react-doctor/src/plugin/rules/tanstack-start/tanstack-start-missing-head-content.ts
import { TANSTACK_ROOT_ROUTE_FILE_PATTERN } from "../../constants/tanstack.js";
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { Rule } from "../../utils/rule.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

const TANSTACK_ROUTER_PACKAGE = "@tanstack/react-router";
const HEAD_CONTENT_COMPONENT_NAME = "HeadContent";
const DOCUMENT_HEAD_ELEMENT_NAME = "head";

const getJsxMemberRootName = (node: EsTreeNodeOfType<"JSXMemberExpression">): string | null => {
  if (isNodeOfType(node.object, "JSXIdentifier")) return node.object.name;
  if (isNodeOfType(node.object, "JSXMemberExpression")) return getJsxMemberRootName(node.object);
  return null;
};

const getJsxMemberPropertyName = (node: EsTreeNodeOfType<"JSXMemberExpression">): string | null => {
  if (isNodeOfType(node.property, "JSXIdentifier")) return node.property.name;
  return null;
};

export const tanstackStartMissingHeadContent = defineRule<Rule>({
  id: "tanstack-start-missing-head-content",
  tags: ["test-noise"],
  requires: ["tanstack-start"],
  severity: "warn",
  recommendation:
    "Add `<HeadContent />` inside `<head>` in your __root route — without it, route `head()` meta tags are silently dropped",
  create: (context: RuleContext) => {
    let hasHeadContentElement = false;
    let hasDocumentHeadElement = false;
    const headContentComponentNames = new Set([HEAD_CONTENT_COMPONENT_NAME]);
    const tanstackRouterNamespaceNames = new Set<string>();

    return {
      ImportDeclaration(node: EsTreeNodeOfType<"ImportDeclaration">) {
        const filename = context.getFilename?.() ?? "";
        const isRootRouteFile = TANSTACK_ROOT_ROUTE_FILE_PATTERN.test(filename);
        if (!isRootRouteFile) return;
        if (node.source.value !== TANSTACK_ROUTER_PACKAGE) return;

        const specifiers = node.specifiers ?? [];
        for (const specifier of specifiers) {
          if (isNodeOfType(specifier, "ImportNamespaceSpecifier")) {
            tanstackRouterNamespaceNames.add(specifier.local.name);
            continue;
          }

          if (!isNodeOfType(specifier, "ImportSpecifier")) continue;
          if (
            !isNodeOfType(specifier.imported, "Identifier") ||
            specifier.imported.name !== HEAD_CONTENT_COMPONENT_NAME
          )
            continue;
          headContentComponentNames.add(specifier.local.name);
        }
      },
      JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
        const filename = context.getFilename?.() ?? "";
        const isRootRouteFile = TANSTACK_ROOT_ROUTE_FILE_PATTERN.test(filename);
        if (!isRootRouteFile) return;

        if (isNodeOfType(node.name, "JSXIdentifier")) {
          if (node.name.name === DOCUMENT_HEAD_ELEMENT_NAME) {
            hasDocumentHeadElement = true;
          }
          if (headContentComponentNames.has(node.name.name)) {
            hasHeadContentElement = true;
          }
          return;
        }

        if (!isNodeOfType(node.name, "JSXMemberExpression")) return;

        const rootName = getJsxMemberRootName(node.name);
        const propertyName = getJsxMemberPropertyName(node.name);
        if (
          rootName &&
          tanstackRouterNamespaceNames.has(rootName) &&
          propertyName === HEAD_CONTENT_COMPONENT_NAME
        ) {
          hasHeadContentElement = true;
        }
      },
      "Program:exit"(programNode: EsTreeNode) {
        const filename = context.getFilename?.() ?? "";
        const isRootRouteFile = TANSTACK_ROOT_ROUTE_FILE_PATTERN.test(filename);
        if (!isRootRouteFile) return;

        if (hasDocumentHeadElement && !hasHeadContentElement) {
          context.report({
            node: programNode,
            message:
              "Root route (__root) without <HeadContent /> — route head() meta tags won't render",
          });
        }
      },
    };
  },
});

File: native/oxlint/rules/tanstack-start-missing-head-content.rs
use oxc_ast::{
    AstKind,
    ast::{
        Expression, ImportDeclarationSpecifier, JSXElementName, JSXMemberExpression,
        JSXMemberExpressionObject, MemberExpression,
    },
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use rustc_hash::FxHashSet;

use crate::{
    context::{ContextHost, LintContext},
    rule::Rule,
};

const TANSTACK_ROUTER_PACKAGE: &str = "@tanstack/react-router";
const HEAD_CONTENT_COMPONENT_NAME: &str = "HeadContent";

#[derive(Debug, Default, Clone)]
pub struct TanstackStartMissingHeadContent;

declare_oxc_lint!(
    /// Require HeadContent in TanStack Start root routes.
    TanstackStartMissingHeadContent,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Require HeadContent in TanStack Start root routes.",
);

impl Rule for TanstackStartMissingHeadContent {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        is_tanstack_root_route_filename(&ctx.file_path().to_string_lossy())
    }

    fn run_once(&self, ctx: &LintContext<'_>) {
        let mut head_content_component_names =
            FxHashSet::from_iter([HEAD_CONTENT_COMPONENT_NAME.to_string()]);
        let mut tanstack_router_namespace_names = FxHashSet::default();

        for node in ctx.nodes().iter() {
            let AstKind::ImportDeclaration(import_declaration) = node.kind() else {
                continue;
            };
            let is_tanstack_router_import = import_declaration.source.value == TANSTACK_ROUTER_PACKAGE;
            for specifier in import_declaration.specifiers.iter().flatten() {
                match specifier {
                    ImportDeclarationSpecifier::ImportNamespaceSpecifier(specifier)
                        if is_tanstack_router_import =>
                    {
                        tanstack_router_namespace_names.insert(specifier.local.name.to_string());
                    }
                    ImportDeclarationSpecifier::ImportSpecifier(specifier)
                        if specifier.imported.name() == HEAD_CONTENT_COMPONENT_NAME =>
                    {
                        head_content_component_names.insert(specifier.local.name.to_string());
                    }
                    _ => {}
                }
            }
        }

        for node in ctx.nodes().iter() {
            let AstKind::VariableDeclarator(declarator) = node.kind() else {
                continue;
            };
            let Some(alias) = declarator.id.get_binding_identifier() else {
                continue;
            };
            let Some(initializer) = &declarator.init else {
                continue;
            };
            match initializer {
                Expression::Identifier(identifier) => {
                    if head_content_component_names.contains(identifier.name.as_str()) {
                        head_content_component_names.insert(alias.name.to_string());
                    }
                    if tanstack_router_namespace_names.contains(identifier.name.as_str()) {
                        tanstack_router_namespace_names.insert(alias.name.to_string());
                    }
                }
                expression => {
                    let Some(member_expression) = expression.as_member_expression() else {
                        continue;
                    };
                    let Some(root_name) = member_expression_root_name(member_expression) else {
                        continue;
                    };
                    if tanstack_router_namespace_names.contains(root_name)
                        && member_expression_identifier_property_name(member_expression)
                            == Some(HEAD_CONTENT_COMPONENT_NAME)
                    {
                        head_content_component_names.insert(alias.name.to_string());
                    }
                }
            }
        }

        let mut has_head_content_element = false;
        let mut has_document_head_element = false;
        let mut has_custom_head_child_element = false;
        for node in ctx.nodes().iter() {
            let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
                continue;
            };
            match &opening_element.name {
                JSXElementName::Identifier(identifier) => {
                    if identifier.name == "head" {
                        has_document_head_element = true;
                    }
                    if head_content_component_names.contains(identifier.name.as_str()) {
                        has_head_content_element = true;
                    }
                }
                JSXElementName::IdentifierReference(identifier) => {
                    if head_content_component_names.contains(identifier.name.as_str()) {
                        has_head_content_element = true;
                    }
                }
                JSXElementName::MemberExpression(member_expression) => {
                    if jsx_member_root_name(member_expression).is_some_and(|root_name| {
                        tanstack_router_namespace_names.contains(root_name)
                    }) && member_expression.property.name == HEAD_CONTENT_COMPONENT_NAME
                    {
                        has_head_content_element = true;
                    }
                }
                _ => {}
            }

            if is_inside_document_head_element(node.id(), ctx)
                && jsx_element_name_is_custom(&opening_element.name)
            {
                has_custom_head_child_element = true;
            }
        }

        if has_document_head_element
            && !has_head_content_element
            && !has_custom_head_child_element
        {
            ctx.diagnostic(
                OxcDiagnostic::warn(
                    "Without <HeadContent /> in the __root route, your route head() meta tags never render.",
                )
                .with_label(program_estree_span(ctx.nodes().program())),
            );
        }
    }
}

fn member_expression_root_name<'a>(member_expression: &'a MemberExpression<'a>) -> Option<&'a str> {
    match member_expression.object() {
        Expression::Identifier(identifier) => Some(identifier.name.as_str()),
        expression => member_expression_root_name(expression.as_member_expression()?),
    }
}

fn jsx_member_root_name<'a>(member_expression: &'a JSXMemberExpression<'a>) -> Option<&'a str> {
    match &member_expression.object {
        JSXMemberExpressionObject::IdentifierReference(identifier) => Some(identifier.name.as_str()),
        JSXMemberExpressionObject::MemberExpression(member_expression) => {
            jsx_member_root_name(member_expression)
        }
        JSXMemberExpressionObject::ThisExpression(_) => None,
    }
}

fn jsx_element_name_is_custom(element_name: &JSXElementName<'_>) -> bool {
    let root_name = match element_name {
        JSXElementName::Identifier(identifier) => Some(identifier.name.as_str()),
        JSXElementName::IdentifierReference(identifier) => Some(identifier.name.as_str()),
        JSXElementName::MemberExpression(member_expression) => jsx_member_root_name(member_expression),
        _ => None,
    };
    root_name
        .and_then(|name| name.chars().next())
        .is_some_and(char::is_uppercase)
}

fn is_inside_document_head_element(
    node_id: oxc_semantic::NodeId,
    ctx: &LintContext<'_>,
) -> bool {
    ctx.nodes().ancestors(node_id).any(|ancestor| {
        let AstKind::JSXElement(element) = ancestor.kind() else {
            return false;
        };
        matches!(
            &element.opening_element.name,
            JSXElementName::Identifier(identifier) if identifier.name == "head"
        )
    })
}
